### PR TITLE
Adding tests for Symfony Httplug Async

### PR DESCRIPTION
**Pull request by XWB**
Fixes https://github.com/php-http/discovery/issues/158

The change introduced in https://github.com/php-http/discovery/commit/366a549f1a9d8f0ac6fc91535a42d7080b560e57 breaks our application.

### CODE REVIEW

**Comment by dbu** [comment]
i don't think this is the correct fix. @Nyholm made a precondition based on an interface. if the scenario does have that interface available but is still not able to instantiate, we need to add additional criteria for instantiation. 

i guess the problem is that checking for the interface can not tell if any implementation is available. but if there is a whole bunch of possible implementations, i dunno how we can determine if there is one of them present... @Nyholm do you have an idea how to handle it? could we tentatively call some factory to see if it finds a solution?

**Comment by Nyholm** [comment]
I love this library but the past few months it feels like we've done too many patch fixes.. 
A few weeks ago I introduced some additional testing to avoid this kind of issue. 

@XWB Can you have a look here: https://github.com/php-http/discovery/blob/1.7.3/.travis.yml#L59-L67 and provide your own test that represent your application to show what is broken?

**Comment by XWB** [comment]
I''l try to make a test. In the meantime, here's the stack trace:

```php
 at Http\Discovery\Exception\DiscoveryFailedException::create(array(object(PuliUnavailableException), object(NoCandidateFoundException), object(NoCandidateFoundException)))
     (vendor/php-http/discovery/src/ClassDiscovery.php:79)
  at Http\Discovery\ClassDiscovery::findOneByType('Psr\\Http\\Message\\ResponseFactoryInterface')
     (vendor/php-http/discovery/src/Psr17FactoryDiscovery.php:53)
  at Http\Discovery\Psr17FactoryDiscovery::findResponseFactory()
     (vendor/symfony/http-client/HttplugClient.php:79)
  at Symfony\Component\HttpClient\HttplugClient->__construct()
     (vendor/php-http/discovery/src/ClassDiscovery.php:203)
  at Http\Discovery\ClassDiscovery::instantiateClass('Symfony\\Component\\HttpClient\\HttplugClient')
     (vendor/php-http/discovery/src/HttpAsyncClientDiscovery.php:40)
  at Http\Discovery\HttpAsyncClientDiscovery::find()
     (vendor/friendsofsymfony/http-cache/src/ProxyClient/HttpDispatcher.php:99)
```

As you can see, it's trying to initialize `Psr\\Http\\Message\\ResponseFactoryInterface` which doesn't work.

**Comment by Nyholm** [comment]
That is not what I see. =)

I see that you are finding the Symfony HTTPlug client. When instantiating `HttplugClient` we need a `Psr\\Http\\Message\\ResponseFactoryInterface` so we are trying to find one, but fails. 

I thought this test would cover your case: 
https://github.com/php-http/discovery/blob/1.7.3/.travis.yml#L67

Are you using version 1.7.3?

**Comment by XWB** [comment]
I do have 1.7.3 installed:

```
$ composer show | grep discovery

php-http/discovery                  1.7.3      Finds installed HTTPlug implementations and PSR-7 message factories
```

And everything works fine when I revert https://github.com/php-http/discovery/commit/366a549f1a9d8f0ac6fc91535a42d7080b560e57 so the `interface_exists` must break something?

**Comment by Nyholm** [comment]
Please add a test similar to the one I linked. 
It is only one row where you specify your dependencies. It will be interesting to see.

**Comment by XWB** [comment]
@Nyholm Done. I added two tests with `friendsofsymfony/http-cache-bundle` and you can see the error in Travis:

```
PHP Fatal error:  Uncaught Http\Discovery\Exception\DiscoveryFailedException: Could not find resource using any discovery strategy. Find more information at http://docs.php-http.org/en/latest/discovery.html#common-errors
 - Puli Factory is not available
 - No valid candidate found using strategy "Http\Discovery\Strategy\CommonClassesStrategy". We tested the following candidates: .
 - No valid candidate found using strategy "Http\Discovery\Strategy\CommonPsr17ClassesStrategy". We tested the following candidates: Nyholm\Psr7\Factory\Psr17Factory, Zend\Diactoros\ResponseFactory, Http\Factory\Diactoros\ResponseFactory, Http\Factory\Guzzle\ResponseFactory, Http\Factory\Slim\ResponseFactory.
 in /home/travis/build/php-http/discovery/build/504b61b/vendor/php-http/discovery/src/Exception/DiscoveryFailedException.php:41
Stack trace:
#0 /home/travis/build/php-http/discovery/build/504b61b/vendor/php-http/discovery/src/ClassDiscovery.php(79): Http\Discovery\Exception\DiscoveryFailedException::create(Array)
#1 /home/travis/build/php-http/ in /home/travis/build/php-http/discovery/build/504b61b/vendor/php-http/discovery/src/ClassDiscovery.php on line 210
```

**Comment by onEXHovia** [comment]
✋ I don’t know if it will be useful. In project we use packages with versions:
```
"friendsofsymfony/http-cache-bundle": "^2.8"
"php-http/guzzle6-adapter": "^1.1"
"php-http/httplug-bundle": "^1.17"
```

`php-http/discovery ` have version 1.7.3, current set packages work correct, but after install `symfony/http-client@v4.4.2` i get the same exception which mentioned in the [comment](https://github.com/php-http/discovery/issues/158#issuecomment-570259279).

File: .travis.yml [hunk 1]
```diff
@@ -65,6 +65,10 @@ matrix:
               - install_test cant-find "Http\Discovery\HttpClientDiscovery::find();" "symfony/http-client:5.* php-http/httplug guzzlehttp/psr7:1.* http-interop/http-factory-guzzle"
               # We should be able to find a client when Symfony is only partly installed and we have guzzle adapter installed
               - install_test will-find "Http\Discovery\HttpClientDiscovery::find();" "symfony/http-client:5.* php-http/guzzle6-adapter php-http/httplug php-http/message-factory guzzlehttp/psr7:1.*"
+              # Test that we find a client with Symfony and Guzzle
+              - install_test will-find "Http\Discovery\HttpClientDiscovery::find();" "friendsofsymfony/http-cache-bundle:2.4.* symfony/http-client:4.* php-http/guzzle6-adapter"
+              # Test that we find an async client with Symfony and Guzzle
+              - install_test will-find "Http\Discovery\HttpAsyncClientDiscovery::find();" "friendsofsymfony/http-cache-bundle:2.4.* symfony/http-client:4.* php-http/guzzle6-adapter"
```
Comment (Nyholm) [inline .travis.yml]: You can replace `friendsofsymfony/http-cache-bundle:2.4.*` with `php-http/client-common:2.* php-http/message:1.8*`

It is the same effect.

**Comment by Nyholm** [comment]
I also added #161. Both will fix this issue. Which one do we prefer?

**Comment by XWB** [comment]
@Nyholm I manually applied your fix in our application and it still fails :(

image

image

**Comment by XWB** [comment]
But https://github.com/php-http/discovery/pull/161 fixes the issue! :)

**Comment by Nyholm** [comment]
Fixed and released https://github.com/php-http/discovery/releases/tag/1.7.4

**Pull request closed without merging**